### DISCUSSION TITLE
Difficulty connecting to MS Azure SQL database

### DISCUSSION THREAD

**Discussion by charlyrock** [sqlalchemy/sqlalchemy#9288] [Usage Questions] [answered]
In my config I try to connect a Cloud Azure Database with sqlalchemy.
My Config File:

```
from flask import Flask, render_template, request, redirect, url_for
from flask_bootstrap import Bootstrap
import sqlalchemy
from sqlalchemy.orm import sessionmaker, scoped_session

current_db = 'mysql'
mysql_engine = sqlalchemy.create_engine('mysql+mysqlconnector://root:root@mysql/db')
mysql_session = scoped_session(sessionmaker(bind=mysql_engine))

postgres_engine = sqlalchemy.create_engine('postgresql://postgres:root@postgres:5432/db')
postgres_session = scoped_session(sessionmaker(bind=postgres_engine))

mssql_engine = sqlalchemy.create_engine('mssql+pymssql://user:password@myconnectstring.database.windows.net:1433/mydb;encrypt=true;trustServerCertificate=false;hostNameInCertificate=*.database.windows.net;loginTimeout=30;')
mssql_session = scoped_session(sessionmaker(bind=mssql_engine))
[...]
```
For MySQL and Postgres it works. But the mssql doesn´t work. It should go to a cloud database. 

If I opened my webserver and try to connect to the database I get the following error:
`sqlalchemy.exc.OperationalError: (pymssql._pymssql.OperationalError) (18456, b"Login failed for user 'user'.DB-Lib error message 20018, severity 14:\nGeneral SQL Server error: Check messages from the SQL Server\nDB-Lib error message 20002, severity 9:\nAdaptive Server connection failed (myconnectionstring.database.windows.net)\nDB-Lib error message 20002, severity 9:\nAdaptive Server connection failed (myconnectionstring.database.windows.net)\n")`

I've tried so much, but nothing helps. And after 3 days try and error i have no more ideas. Can you help?

**Comment by zzzeek** [comment, marked as answer]
you need to use ODBC to connect to an azure cloud DB.   use [pyodbc](https://pypi.org/project/pyodbc/) with the [Microsoft SQL Server ODBC drivers](https://learn.microsoft.com/en-us/sql/connect/odbc/download-odbc-driver-for-sql-server?view=sql-server-ver16) , pyodbc connect examples with several for azure are at https://docs.sqlalchemy.org/en/20/dialects/mssql.html#connecting-to-pyodbc

**Reply by charlyrock** [reply]
Thanks. I´ve tried it before. The problem is, I use Python3.8 and on [Pypi-Website](https://pypi.org/project/pyodbc/) stands that:
IMPORTANT: Python 2.7 support is being ended. The pyodbc 4.x versions will be the last to support Python 2.7. The pyodbc 5.x versions will only support Python 3.7 and above.

But acutally it´s no newer version as 4.x. The pyodbc version 5 I can´t find anywhere. 
That´s why I´ve tried the alternative pymssql. Any other ideas?

**Reply by zzzeek** [reply]
any version of pyodbc will work with python 3.8, the 4.x series or whatever should be fine.

**Reply by charlyrock** [reply]
Okay. Thanks. I´ve tried it again.
But now I got this error:
`myapp_1       |     dbapi = dialect_cls.dbapi(**dbapi_args)
myapp_1       |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/connectors/pyodbc.py", line 38, in dbapi
myapp_1       |     return __import__("pyodbc")
myapp_1       | ImportError: libodbc.so.2: cannot open shared object file: No such file or directory
`

But I have installed libodbc and libodbc-dev

Here is my Dockerfile:

```
FROM debian:buster-slim
RUN apt-get update \
 && apt-get -y upgrade
RUN apt-get install -y python3-pip

RUN apt-get install -y wget curl lsb-release ca-certificates apt-transport-https software-properties-common gnupg2 \
 && wget -qO /etc/apt/trusted.gpg.d/php.gpg https://packages.sury.org/php/apt.gpg \
 && echo "deb https://packages.sury.org/php/ $(lsb_release -sc) main" > /etc/apt/sources.list.d/sury-php.list 

RUN apt-get update \
 && apt-get install -y php8.0 php8.0-dev php8.0-xml php8.0-odbc php8.0-sybase

RUN apt-get -y install unixodbc-dev

[...]
RUN curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add - \
 && curl https://packages.microsoft.com/config/debian/9/prod.list > /etc/apt/sources.list.d/mssql-release.list

RUN apt-get update \
 && ACCEPT_EULA=Y apt-get -y install msodbcsql17 

[...]

FROM php:8.0-fpm

RUN apt-get update
RUN apt-get -y install gcc make autoconf libc-dev pkg-config libzip-dev

RUN apt-get install -y --no-install-recommends \
	git \
	libmemcached-dev \
	libz-dev \
	libpq-dev \
	libssl-dev libssl-doc libsasl2-dev \
	libmcrypt-dev \
	libxml2-dev \
	zlib1g-dev libicu-dev g++ \
	libldap2-dev libbz2-dev \
	curl libcurl4-openssl-dev \
	libgmp-dev firebird-dev libib-util \
	re2c libpng++-dev \
	libmagick++-dev \
	libmagickwand-dev \
	zlib1g-dev libgd-dev \
	libtidy-dev libxslt1-dev libmagic-dev libexif-dev file \
	sqlite3 libsqlite3-dev libxslt-dev \
	libmhash2 libmhash-dev libc-client-dev libkrb5-dev libssh2-1-dev \
	unzip libpcre3 libpcre3-dev \
	poppler-utils ghostscript libmagickwand-6.q16-dev libsnmp-dev libedit-dev libreadline6-dev libsodium-dev \
	freetds-bin freetds-common libct4 tdsodbc libreadline-dev librecode-dev libpspell-dev libonig-dev unixodbc unixodbc-dev

RUN apt-get install -y freetds-dev libsybdb5 \
	&& ln -s /usr/lib/x86_64-linux-gnu/libsybdb.a /usr/lib/libsybdb.a

RUN docker-php-ext-configure imap --with-kerberos --with-imap-ssl && \
	docker-php-ext-install imap iconv

RUN docker-php-ext-install mysqli pdo pdo_mysql

RUN docker-php-ext-configure pdo_odbc --with-pdo-odbc=unixODBC,/usr
RUN docker-php-ext-install pdo_odbc

RUN docker-php-ext-configure pdo_dblib --with-libdir=lib/ && \
 docker-php-ext-install pdo_dblib


FROM python:3.8-buster

ENV FLASK_APP=app.py \
    FLASK_ENV=development \
    FLASK_DEBUG=1

RUN mkdir /app
RUN pip install pipenv
RUN pip install pyodbc


WORKDIR /app
EXPOSE 80
```

**Reply by zzzeek** [reply]
@gordthompson  is there an LD_LIBRARY_PATH needed for this ?

**Reply by charlyrock** [reply]
I tried it with:
`RUN export LIBRARY_PATH=$LIBARY_PATH:/usr/lib/x86_x64-linux-gnu/`

and:
```
ENV PATH=${PATH}:/usr/lib/x86_x64-linux-gnu/
ENV PATH=$PATH:/usr/lib/x86_x64-linux-gnu/
```

but the error remains. 
Maybe I must configure it at a other part?
I wrote it a line above "RUN pip install pyodbc"

**Reply by gordthompson** [reply]
> @gordthompson is there an LD_LIBRARY_PATH needed for this ?

No. One suggestion would be to omit

```
RUN apt-get -y install unixodbc-dev
```

because

```
ACCEPT_EULA=Y apt-get -y install msodbcsql17
```

installs unixodbc as a dependency and gets it from Microsoft's repository instead of Debian's. (Debian's version might be considerably older.)

**Reply by charlyrock** [reply]
Thank you. I´ve tried it without 
`RUN apt-get -y install unixodbc-dev`

But now my install of pdo_odbc fails because it use unixODBC. 
`configure: error: Cannot find header file(s) for pdo_odbc
The command '/bin/sh -c docker-php-ext-configure pdo_odbc --with-pdo-odbc=unixODBC,/usr' returned a non-zero code: 1`

But for the application I need pdo_odbc. 
Can I give the path to msodbcsql or any alternatives?

**Reply by gordthompson** [reply]
Oh, okay. If pdo_odbc needs unixodbc-dev then I would install msodbcsql17 first, followed by unixodbc-dev, and then pdo_odbc.

**Reply by charlyrock** [reply]
Thanks so much. But unfortunately that doesn't result in any change and the error remains.
`
app_1       | Traceback (most recent call last):
app_1       |   File "app.py", line 13, in <module>
app_1       |     mssql_engine = sqlalchemy.create_engine('mssql+pyodbc://user:password@myconnectionstrin.database.windows.net:1433/dbs1;encrypt=true;trustServerCertificate=false;hostNameInCertificate=*.database.windows.net;loginTimeout=30;')
app_1       |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/__init__.py", line 500, in create_engine
app_1       |     return strategy.create(*args, **kwargs)
app_1       |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/strategies.py", line 87, in create
app_1       |     dbapi = dialect_cls.dbapi(**dbapi_args)
app_1       |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/connectors/pyodbc.py", line 38, in dbapi
app_1       |     return __import__("pyodbc")
app_1       | ImportError: libodbc.so.2: cannot open shared object file: No such file or directory
`

**Reply by CaselIT** [reply]
The documentation webside of microsoft has this suggested script to install pyodbc:
https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server?redirectedfrom=MSDN&view=sql-server-ver16&tabs=debian18-install%2Cdebian17-install%2Cdebian8-install%2Credhat7-13-install%2Crhel7-offline#17

that seems compatible with what you are doing in your docker file.

My suggestion would be start with a barebone docker file that has only this library and try to see if you can connect with it, maybe using the command line tools like `sqlcmd`.
If it works then start re-adding the things in your current docker file until you find what's failing it.

**Reply by gordthompson** [reply]
I agree. Something else that you are installing is breaking the unixODBC configuration. This MCVE works:

```
gord@vm1720:~/mydockerbuild$ cat my_app.py 
import pyodbc

cnxn = pyodbc.connect(
    "Driver=ODBC Driver 17 for SQL Server;"
    "Server=192.168.0.199;"
    "UID=scott;"
    "PWD=tiger^5HHH;"
    "Database=test;"
)
crsr = cnxn.cursor()
print(crsr.execute("SELECT '>>> Connect succeeded.'").fetchval())
crsr.close()
cnxn.close()
gord@vm1720:~/mydockerbuild$ # ----------
gord@vm1720:~/mydockerbuild$ cat my_dockerfile 
FROM python:3.8
WORKDIR /app
COPY my_app.py my_app.py
RUN apt-get update && apt-get -y install curl
RUN curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add - && curl https://packages.microsoft.com/config/debian/10/prod.list > /etc/apt/sources.list.d/mssql-release.list
RUN apt-get update && ACCEPT_EULA=Y apt-get -y install msodbcsql17
RUN pip install pyodbc
RUN python my_app.py
gord@vm1720:~/mydockerbuild$ # ----------
gord@vm1720:~/mydockerbuild$ docker build --progress=plain --no-cache ~/mydockerbuild -f my_dockerfile -t sqla_disc_9288
#1 [internal] load .dockerignore
#1 transferring context: 2B 0.0s done
#1 DONE 0.1s

#2 [internal] load build definition from my_dockerfile
#2 transferring dockerfile: 439B 0.0s done
#2 DONE 0.2s

#3 [internal] load metadata for docker.io/library/python:3.8
#3 DONE 0.0s

#4 [1/8] FROM docker.io/library/python:3.8
#4 DONE 0.0s

#5 [2/8] WORKDIR /app
#5 CACHED

#6 [internal] load build context
#6 transferring context: 329B 0.0s done
#6 DONE 0.1s

#7 [3/8] COPY my_app.py my_app.py
#7 DONE 0.3s

#8 [4/8] RUN apt-get update && apt-get -y install curl
#8 …
#8 DONE 34.3s

#9 [5/8] RUN curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add - && curl https://packages.microsoft.com/config/debian/10/prod.list > /etc/apt/sources.list.d/mssql-release.list
#9 …
#9 DONE 7.1s

#10 [6/8] RUN apt-get update && ACCEPT_EULA=Y apt-get -y install msodbcsql17
#10 …
#10 DONE 25.3s

#11 [7/8] RUN pip install pyodbc
#11 12.16 Collecting pyodbc
#11 12.29   Downloading pyodbc-4.0.35-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (333 kB)
#11 12.79 Installing collected packages: pyodbc
#11 12.92 Successfully installed pyodbc-4.0.35
#11 DONE 15.1s

#12 [8/8] RUN python my_app.py
#12 2.597 >>> Connect succeeded.
#12 DONE 3.1s

#13 exporting to image
#13 …
#13 DONE 2.3s
```

**Reply by charlyrock** [reply]
Thank you both sooo much. 
I was able to get a little bit further in my environment. Since it is passed on to a website and reused, I use sqlalchemy. The connection also seems to work. And the errors regarding pyodbc etc could be fixed. However, this error always occurs. Unfortunately, a few searches and repeated switching and specifying the driver in the connection string did not lead to any change.
`
sqlalchemy.exc.InterfaceError: (pyodbc.InterfaceError) ('IM002', '[IM002] [unixODBC][Driver Manager]Data source name not found and no default driver specified (0) (SQLDriverConnect)')`

**Reply by gordthompson** [reply]
Check your connection URL. It should be constructed something like this:

```python
connection_url = (
    "mssql+pyodbc://scott:tiger^5HHH@192.168.0.199/test"
    "?driver=ODBC Driver 18 for SQL Server&TrustServerCertificate=yes"
)
```

In this case the database name is 'test'. Note that the entry immediately following the database name begins with a question mark (`?`) and subsequent key/value pairs begin with an ampersand (`&`), not a semicolon (`;`).

**Reply by charlyrock** [reply]
Thank you all very much for your help. That helps so much and at last a env variable must be set and now it works. 
Thank you so sooo much!!!
